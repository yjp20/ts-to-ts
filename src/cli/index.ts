#!/usr/bin/env node

import { runConvert } from "./commands/convert.js";

runConvert(process.argv.slice(2)).catch((error: Error) => {
  console.error(error);
  process.exit(1);
});
