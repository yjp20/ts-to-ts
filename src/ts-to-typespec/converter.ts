import {
  Project,
  SourceFile,
  Type,
  TypeFormatFlags,
  Node,
  SyntaxKind,
} from "ts-morph";

type ConversionResult =
  | {
      kind: "error";
      file: string;
      error: Error;
    }
  | {
      kind: "success";
      file: string;
      result: string;
    };

export function convertProject(project: Project, files: string[]): string {
  const conversions = files.map((file): ConversionResult => {
    const sourceFile = project.getSourceFile(file);
    if (!sourceFile) {
      return {
        kind: "error",
        file,
        error: new Error(`Source file ${JSON.stringify(file)}`),
      };
    }

    return {
      kind: "success",
      file,
      result: convertSourceFile(sourceFile),
    };
  });

  return conversions
    .filter((conversion) => conversion.kind === "success")
    .map((conversion) => `// ${conversion.file}\n${conversion.result}`)
    .join("\n\n");
}

export function convertSourceFile(sourceFile: SourceFile): string {
  const nodes = [
    ...sourceFile.getChildrenOfKind(SyntaxKind.TypeAliasDeclaration),
    ...sourceFile.getChildrenOfKind(SyntaxKind.InterfaceDeclaration),
  ];

  const convertibleNodes = nodes.filter((node) => {
    const leadingComments = node.getLeadingCommentRanges();
    return leadingComments.some((comment) =>
      comment.getText().includes("@model")
    );
  });

  return convertibleNodes
    .map((node) => convertTypeDefinition(node))
    .filter((text) => text.length > 0)
    .join("\n\n");
}

export function convertTypeDefinition(node: Node): string {
  if (Node.isTypeAliasDeclaration(node)) {
    const name = node.getName();
    const type = node.getType();
    const typeText = convertType(type);

    return `alias ${name} = ${typeText};`;
  } else if (Node.isInterfaceDeclaration(node)) {
    const name = node.getName();
    const type = node.getType();
    const typeText = convertType(type);

    return `model ${name} ${typeText};`;
  }

  return ``;
}

function mustBeModel(type: Type): boolean {
  return type.isObject();
}

export function convertType(type: Type): string {
  if (type.isString()) {
    return "string";
  }
  if (type.isTemplateLiteral()) {

  }
  if (type.isNumber()) {
    return "float64";
  }
  if (type.isBoolean()) {
    return "boolean";
  }
  if (type.isArray()) {
    // Handle tuple types
    const elementType = type.getArrayElementType()!;
    return `${convertType(elementType)}[]`;
  }

  if (type.isTuple()) {
    const tupleTypes = type.getTupleElements();
    return `[${tupleTypes.map((t) => convertType(t)).join(", ")}]`;
  }

  if (type.isObject()) {
    const symbol = type.getSymbol();

    // Handle generic types
    if (type.getTypeArguments().length > 0) {
      const baseType = symbol?.getName() || "";
      const typeArgs = type.getTypeArguments().map((t) => convertType(t));

      // Special case for built-in generics
      if (baseType === "Array") {
        return `${typeArgs[0]}[]`;
      }

      if (baseType === "Record") {
        return `Record<${typeArgs.join(", ")}>`;
      }

      return `${baseType}<${typeArgs.join(", ")}>`;
    }

    const properties = type.getProperties();
    const propertyStrings = properties.map((prop) => {
      const type = convertType(
        prop.getTypeAtLocation(prop.getValueDeclaration()!)
      );
      return `${prop.getName()}${prop.isOptional() ? "?" : ""}: ${type}`;
    });

    if (properties.length === 0) {
      return "Record<never, never>";
    }

    return `{\n  ${propertyStrings.join(",\n  ")}\n}`;
  }

  if (type.isUnion()) {
    return type
      .getUnionTypes() //
      .map(convertType)
      .join(" | ");
  }

  if (type.isIntersection()) {
    const types = type.getIntersectionTypes();
    return `{\n  ...${types.map((t) => convertType(t)).join(",\n  ...")}\n}`;
  }

  // Default case - use the type's text representation with proper formatting
  const typeText = type.getText(
    undefined,
    TypeFormatFlags.NoTruncation |
      TypeFormatFlags.WriteClassExpressionAsTypeLiteral
  );

  return typeText;
}
