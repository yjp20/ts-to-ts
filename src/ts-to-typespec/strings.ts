function DumbCombinator<T>(
  values: Array<string | number | LazyString<T>>,
  toString: (rendered: string[]) => string
): LazyString<T> {
  return {
    attach: (scope) => {
      for (const value of values) {
        if (typeof value === "object" && "attach" in value && value.attach) {
          value.attach(scope);
        }
      }
    },

    toString: (scope) => {
      return toString(
        values.map((value) => {
          if (typeof value === "object" && "toString" in value) {
            return value.toString(scope);
          }
          return String(value);
        })
      );
    },
  };
}

export function lazy<T>(
  strings: TemplateStringsArray,
  ...values: Array<string | number | LazyString<T>>
): LazyString<T> {
  return DumbCombinator(values, (rendered) =>
    String.raw({ raw: strings.raw }, ...rendered)
  );
}

export function Indent<T>(
  text: string | LazyString<T>,
  level = 1
): LazyString<T> {
  const spaces = "  ".repeat(level);
  return DumbCombinator([text], (rendered) =>
    rendered
      .flatMap((part) => part.split("\n"))
      .map((line) => (line ? spaces + line : line))
      .join("\n")
  );
}

export function Stanzas<T>(...values: LazyString<T>[]): LazyString<T> {
  return DumbCombinator(values, (rendered) => rendered.join("\n\n"));
}

export function Lines<T>(...values: LazyString<T>[]): LazyString<T> {
  return DumbCombinator(values, (rendered) => rendered.join("\n"));
}

export function Concat<T>(...values: LazyString<T>[]): LazyString<T> {
  return DumbCombinator(values, (rendered) => rendered.join(""));
}

export function Union<T>(...values: LazyString<T>[]): LazyString<T> {
  return DumbCombinator(values, (rendered) => rendered.join(" | "));
}

export function Commas<T>(...values: LazyString<T>[]): LazyString<T> {
  return DumbCombinator(values, (rendered) => rendered.join(","));
}

export type LazyString<Scope> = {
  attach: (scope: Scope) => void;
  toString(scope?: Scope): string;
};
