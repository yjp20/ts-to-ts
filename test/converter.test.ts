import { TypeScriptToTypeSpecConverter } from "../src/ts-to-typespec";
import * as fs from "fs";
import * as path from "path";
import { NodeHost, compile } from "@typespec/compiler";

describe("TypeScriptToTypeSpecConverter", () => {
  let converter: TypeScriptToTypeSpecConverter;
  let tempDir: string;

  beforeEach(() => {
    converter = new TypeScriptToTypeSpecConverter();
    tempDir = path
      .join(path.dirname(import.meta.url), "../temp")
      .replace("file:", "");
    if (!fs.existsSync(tempDir)) {
      fs.mkdirSync(tempDir);
    }
  });

  afterEach(() => {
    if (fs.existsSync(tempDir)) {
      fs.rmSync(tempDir, { recursive: true });
    }
  });

  // Dynamically load and run tests for all fixtures
  const fixturesDir = path
    .join(path.dirname(import.meta.url), "fixtures")
    .replace("file:", "");
  const fixtures = fs
    .readdirSync(fixturesDir)
    .filter((file) => file.endsWith(".ts"));

  for (const fixture of fixtures) {
    const testName = path.basename(fixture, ".ts");

    test(testName, async () => {
      // Read fixture file
      const fileContent = fs.readFileSync(
        path.join(fixturesDir, fixture),
        "utf-8"
      );

      // Convert to TypeSpec
      const typespecCode = converter.convertTypeToTypeSpec(fileContent);

      // Write TypeSpec code to temp directory
      const tempFile = path.join(tempDir, "main.tsp");
      fs.writeFileSync(tempFile, typespecCode);
      fs.writeFileSync(
        path.join(tempDir, "package.json"),
        JSON.stringify({ name: "test", main: "main.tsp" })
      );

      // Verify TypeSpec compilation
      const host = new NodeHost();
      expect(
        "\n<<<<<<<<<<< typescript <<<<<<<<<<<<\n" +
          fileContent.trim() +
          "\n===================================\n" +
          typespecCode.trim() +
          "\n>>>>>>>>>>> typespec >>>>>>>>>>>>>>"
      ).toMatchSnapshot(testName);

      const program = await compile(host, ".");
      const errors = program.diagnostics.filter((d) => d.severity === "error");
      expect(errors).toHaveLength(0);
    });
  }
});
