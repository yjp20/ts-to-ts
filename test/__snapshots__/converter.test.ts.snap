// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`convert annotated-types: annotated-types 1`] = `
"
<<<<<<<<<<< typescript <<<<<<<<<<<<
/** @model */
type ConvertMe = string;

type SkipMe = number;

/** @model */
interface UserInfo {
  name: string;
  age: number;
}

interface IgnoreMe {
  data: string;
}
===================================
// annotated-types.ts
alias ConvertMe = string;

model UserInfo {
  name: ConvertMe,
  age: float64
};
>>>>>>>>>>> typespec >>>>>>>>>>>>>>"
`;

exports[`convert array-type: array-type 1`] = `
"
<<<<<<<<<<< typescript <<<<<<<<<<<<
/** @model */
type StringArray = string[];
===================================
// array-type.ts
alias StringArray<T> = string[];
>>>>>>>>>>> typespec >>>>>>>>>>>>>>"
`;

exports[`convert complex-type: complex-type 1`] = `
"
<<<<<<<<<<< typescript <<<<<<<<<<<<
/** @model */
interface User {
  id: number;
  name: string;
  roles: string[];
  status: "active" | "inactive";
  profile: {
    email: string;
    age?: number;
  };
}
===================================
// complex-type.ts
model User {
  id: float64,
  name: string,
  roles: string[],
  status: "active" | "inactive",
  profile: {
    email: string,
    age?: float64
  }
};
>>>>>>>>>>> typespec >>>>>>>>>>>>>>"
`;

exports[`convert custom-model-names: custom-model-names 1`] = `
"
<<<<<<<<<<< typescript <<<<<<<<<<<<
/** @model CustomGreeting */
type Greeting = string;

/** @model UserProfile */
interface User {
  name: string;
  age: number;
}

/** @model AdminData */
type Admin = User & {
  permissions: string[];
};
===================================
// custom-model-names.ts
alias CustomGreeting = string;

model AdminData {
  name: CustomGreeting,
  age: float64,
  permissions: CustomGreeting[]
};

model UserProfile {
  name: CustomGreeting,
  age: float64
};
>>>>>>>>>>> typespec >>>>>>>>>>>>>>"
`;

exports[`convert generic-types: generic-types 1`] = `
"
<<<<<<<<<<< typescript <<<<<<<<<<<<
/** @model */
interface Container<T> {
  value: T;
  metadata?: Record<string, unknown>;
}

/** @model */
type StringContainer = Container<string>;
/** @model */
type NumberList = Array<number>;
/** @model */
type Dictionary = Record<string, any>;
===================================
// generic-types.ts
model StringContainer<T> {
  value: string,
  metadata?: Record<never, never>
};

alias NumberList<T> = float64[];

alias Dictionary = Record<never, never>;

model Container<T> {
  value: T,
  metadata?: Record<never, never>
};
>>>>>>>>>>> typespec >>>>>>>>>>>>>>"
`;

exports[`convert interface-primitives: interface-primitives 1`] = `
"
<<<<<<<<<<< typescript <<<<<<<<<<<<
/** @model */
interface User {
  id: number;
  name: string;
  isActive: boolean;
}
===================================
// interface-primitives.ts
model User {
  id: float64,
  name: string,
  isActive: boolean
};
>>>>>>>>>>> typespec >>>>>>>>>>>>>>"
`;

exports[`convert intersection-type: intersection-type 1`] = `
"
<<<<<<<<<<< typescript <<<<<<<<<<<<
/** @model */
interface Name {
  first: string;
  last: string;
}

/** @model */
interface Age {
  age: number;
}

/** @model */
type Person = Name & Age;
===================================
// intersection-type.ts
model Person {
  first: string,
  last: string,
  age: float64
};

model Name {
  first: string,
  last: string
};

model Age {
  age: float64
};
>>>>>>>>>>> typespec >>>>>>>>>>>>>>"
`;

exports[`convert literal-types: literal-types 1`] = `
"
<<<<<<<<<<< typescript <<<<<<<<<<<<
/** @model */
type Direction = "north" | "south" | "east" | "west";
type HttpStatus = 200 | 404 | 500;
===================================
// literal-types.ts
alias Direction = "north" | "south" | "east" | "west";
>>>>>>>>>>> typespec >>>>>>>>>>>>>>"
`;

exports[`convert nested-intersection: nested-intersection 1`] = `
"
<<<<<<<<<<< typescript <<<<<<<<<<<<
/** @model */
interface Address {
  street: string;
  city: string;
}

interface Contact {
  email: string;
  phone?: string;
}

interface PersonDetails {
  name: string;
  age: number;
}

type CompleteProfile = PersonDetails & Contact & Address;
===================================
// nested-intersection.ts
model Address {
  street: string,
  city: string
};
>>>>>>>>>>> typespec >>>>>>>>>>>>>>"
`;

exports[`convert optional-properties: optional-properties 1`] = `
"
<<<<<<<<<<< typescript <<<<<<<<<<<<
/** @model */
interface User {
  id: number;
  email?: string;
  nickname?: string | null;
}
===================================
// optional-properties.ts
model User {
  id: float64,
  email?: string,
  nickname?: string
};
>>>>>>>>>>> typespec >>>>>>>>>>>>>>"
`;

exports[`convert simple-type: simple-type 1`] = `
"
<<<<<<<<<<< typescript <<<<<<<<<<<<
/** @model */
type SimpleType = string;
===================================
// simple-type.ts
alias SimpleType = string;
>>>>>>>>>>> typespec >>>>>>>>>>>>>>"
`;

exports[`convert template-literal-types: template-literal-types 1`] = `
"
<<<<<<<<<<< typescript <<<<<<<<<<<<
/** @model */
type Greeting = \`hello \${string}\`;
/** @model */
type EventName = \`\${string}:click\` | \`\${string}:hover\`;
/** @model */
type Color = \`#\${string}\`;
===================================
// template-literal-types.ts
alias Greeting = "hello \${string}";

alias EventName = "\${string}:click" | "\${string}:hover";

alias Color = "#\${string}";
>>>>>>>>>>> typespec >>>>>>>>>>>>>>"
`;

exports[`convert tuple-types: tuple-types 1`] = `
"
<<<<<<<<<<< typescript <<<<<<<<<<<<
/** @model */
type Point = [number, number];
/** @model */
type NamedPoint = [string, number, number];
/** @model */
type MixedTuple = [string, number, boolean?];
===================================
// tuple-types.ts
alias Point = [float64, float64];

alias NamedPoint = [string, float64, float64];

alias MixedTuple = [string, float64, boolean?];
>>>>>>>>>>> typespec >>>>>>>>>>>>>>"
`;

exports[`convert union-type: union-type 1`] = `
"
<<<<<<<<<<< typescript <<<<<<<<<<<<
/** @model */
type Status = "active" | "inactive";
===================================
// union-type.ts
alias Status = "active" | "inactive";
>>>>>>>>>>> typespec >>>>>>>>>>>>>>"
`;
